zenoh-plugin-ros2dds/src/node_info.rs

```diff
@@ -1785,14 +1803,24 @@ impl NodeInfo {
     }
 
     // Remove a DDS Writer possibly used by this node, and returns an UndiscoveredX event if
-    // this Writer was used by some Subscription, Service or Action
+    // this Writer was used by some Publication, Service or Action
     pub fn remove_writer(&mut self, writer: &Gid) -> Option<ROS2DiscoveryEvent> {
         use ROS2DiscoveryEvent::*;
         let node_fullname = self.fullname().to_string();
-        if let Some((name, _)) = self.msg_pub.iter().find(|(_, v)| v.writer == *writer) {
+        // Search in Publishers list if one is using the writer
+        if let Some(name) = self.msg_pub.iter_mut().find_map(|(name, publ)| {
+            if publ.writers.remove(writer) && publ.writers.is_empty() {
+                // found Publisher using the writer: remove the writer from list
+                // and if the list is empty return the Publisher name to "undiscover" it
+                Some(name.clone())
+            } else {
+                None
+            }
+        }) {
+            // Return undiscovery event for this Publisher, since all its DDS Writer have been undiscovered
             return Some(UndiscoveredMsgPub(
                 node_fullname,
-                self.msg_pub.remove(&name.clone()).unwrap(),
+                self.msg_pub.remove(&name).unwrap(),
```

Comment:
Based on https://github.com/eclipse-zenoh/roadmap/discussions/107, `unwrap()` could be avoided here. However, multiple `unwrap()` are already present in the code, which should be IMHO removed. I'd would leave that to a separate PR.

Reply:
For this specific `unwrap()`, the panic should never happen here. The `name` has just been found in `self.msg_pub` hashmap (see call to `find_map()`). And this code is not async.

Do you think `.expect("cannot happen")` better than `unwrap()` in such case ?